### PR TITLE
8296155: Simplify CDS heap region address calculation

### PR DESCRIPTION
**Pull request by iklam**
Rewrite and simplify the convoluted logic used to calculate the runtime location of a CDS heap region. Now it's as simple as:

```
FileMapRegion::dumptime_base() + ArchiveHeapLoader::runtime_delta();
```

See `FileMapInfo::heap_region_runtime_start_address` for an example.

Also removed dead code:

- `FileMapInfo::is_in_shared_region`
- `MetaspaceShared::is_in_shared_region`
- `FileMapInfo::stop_sharing_and_unmap`

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8296155](https://bugs.openjdk.org/browse/JDK-8296155): Simplify CDS heap region address calculation

### Reviewers
 * [Coleen Phillimore](https://openjdk.org/census#coleenp) (@coleenp - **Reviewer**) ⚠️ Review applies to [5cbbe3a5](https://git.openjdk.org/jdk/pull/10923/files/5cbbe3a5153224cc59917c16d614eb14b671c712)
 * [Matias Saavedra Silva](https://openjdk.org/census#matsaave) (@matias9927 - Author) ⚠️ Review applies to [5cbbe3a5](https://git.openjdk.org/jdk/pull/10923/files/5cbbe3a5153224cc59917c16d614eb14b671c712)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/10923/head:pull/10923` \
`$ git checkout pull/10923`

Update a local copy of the PR: \
`$ git checkout pull/10923` \
`$ git pull https://git.openjdk.org/jdk pull/10923/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 10923`

View PR using the GUI difftool: \
`$ git pr show -t 10923`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/10923.diff">https://git.openjdk.org/jdk/pull/10923.diff</a>

</details>

### CODE REVIEW

File: src/hotspot/share/cds/filemap.cpp [hunk 1]
```diff
@@ -1549,11 +1550,22 @@ size_t FileMapRegion::used_aligned() const {
   return align_up(used(), MetaspaceShared::core_region_alignment());
 }
 
-void FileMapRegion::init(int region_index, size_t mapping_offset, size_t size, bool read_only,
+void FileMapRegion::init(int region_index, size_t mapping_offset, char* dumptime_base, size_t size, bool read_only,
                          bool allow_exec, int crc) {
   _is_heap_region = HeapShared::is_heap_region(region_index);
   _is_bitmap_region = (region_index == MetaspaceShared::bm);
+
+  if (_is_heap_region) {
```
Comment (coleenp) [inline src/hotspot/share/cds/filemap.cpp:1558]: Explain please.  Only heap region has a non-null dumptime base?

File: src/hotspot/share/cds/filemap.cpp [hunk 2]
```diff
@@ -1628,14 +1641,7 @@ void FileMapInfo::write_region(int region, char* base, size_t size,
...
     requested_base = base;
-    if (UseCompressedOops) {
-      mapping_offset = (size_t)CompressedOops::encode_not_null(cast_to_oop(base));
-    } else {
-#if INCLUDE_G1GC
-      mapping_offset = requested_base - (char*)G1CollectedHeap::heap()->reserved().start();
-#endif
-    }
-    assert(mapping_offset == (size_t)(uint32_t)mapping_offset, "must be 32-bit only");
+    dumptime_base = base; // This is the rumtime address of the base (lower end) of this heap region.
```
Comment (coleenp) [inline src/hotspot/share/cds/filemap.cpp:1644]: typo: rumtime.

**Review by coleenp — APPROVED** [review]
This looks like a simplification.  is_in_shared_metaspace() is the query for whether something is in shared region now?

File: src/hotspot/share/cds/filemap.cpp (hunk 1 above)
Comment (coleenp) [reply src/hotspot/share/cds/filemap.cpp:1558]: ie add a comment here too.

**Review by matias9927 — APPROVED** [review]
It seems much cleaner and a bit more intuitive now. The removal of dead code is very nice as well. Looks good to me!

File: src/hotspot/share/cds/filemap.cpp (hunk 2 above)
Comment (iklam) [reply src/hotspot/share/cds/filemap.cpp]: Fixed.

File: src/hotspot/share/cds/filemap.cpp (hunk 1 above)
Comment (iklam) [reply src/hotspot/share/cds/filemap.cpp:1560]: I added comments here. There are also comments in cds.h that explains what these fields should be for different types of regions.

**Comment by iklam** [comment]
After further consideration, this PR is not the right way to do it. If we remember the absolute address of the heap regions during dump time, it may cause unnecessary relocation at run time.

We should keep using CDSFileMapRegion::_mapping_offset to encode the requested address for the heap regions. However, the code should be cleaned up and documentation shoukld be corrected. 

I'll open a new RFE. Closing this RFE as will-not-fix.

```
  size_t  _mapping_offset;    // This encodes the requested address for this region to be mapped at runtime.
                              // However, the JVM may choose to map at an alternative location (e.g., for ASLR,
                              // or to adapt to the available ranges in the Java heap range).
                              // - For an RO/RW region, the requested address is:
                              //     FileMapHeader::requested_base_address() + _mapping_offset
                              // - For a heap region, the requested address is:
                              //     +UseCompressedOops: CompressedOops::decode_raw_not_null(_mapping_offset)
                              //     -UseCompressedOops: FileMapHeader::heap_begin() + _mapping_offset
                              // - For heap regions, the _mapping_offset is always zero. The runtime address
                              //   is picked by the OS.
```

**Pull request closed without merging**